File: db/golang/internal/samuraidb/samurai_db.go
package samuraidb

import (
	im "samurai-db/internal/index_manager"
	sm "samurai-db/internal/segment_manager"
)

//type SegmentManager interface {
//	Set(key string, data any) (int64, int64, error)
//	Get(offset, segment int64) (map[string]any, error)
//}

type SamuraiDB struct {
	segmentManager *sm.SegmentManager // use interface SegmentManager
	indexManager   *im.IndexManager
}

func NewSamuraiDB(fileAdapter *sm.SegmentManager, indexManager *im.IndexManager) *SamuraiDB {
	return &SamuraiDB{segmentManager: fileAdapter, indexManager: indexManager}
}

func (db *SamuraiDB) Init() error {
	return db.indexManager.Init()
}

func (db *SamuraiDB) Set(key string, data interface{}) error {
	offset, segment, err := db.segmentManager.Set(key, data)
	if err != nil {
		return err
	}
	return db.indexManager.SetOffset(key, offset, segment)
}

func (db *SamuraiDB) Get(key string) (map[string]any, error) {
	indexData, exists := db.indexManager.GetOffset(key)
	if !exists {
		return nil, nil
	}

	get, err := db.segmentManager.Get(indexData.Offset, indexData.Segment)
	if err != nil {
		return nil, err
	}

	return get, nil
}

File: db/nodejs/v1/src/core/samuraidb.ts
import { FileAdapter } from './file.adapter';
import {IndexManager} from "./index-manager";
import {SegmentManager} from "./segment-manager";
import {CompactionManager} from "./compaction-manager";

class SamuraiDB {
    /**
     *
     * @param indexManager
     */
    constructor(protected segmentManager: SegmentManager,
                protected indexManager: IndexManager,
                protected compactionManager: CompactionManager) {}

    async init() {
        await this.indexManager.init()
        this.runCompactionInterval()
    }

    runCompactionInterval() {
        setInterval( async () => {
            await this.compactionManager.compactSegments();
            this.runCompactionInterval();
            console.log("COMPACTION FINISHED")
        }, 10 * 1000)
    }

    async set(key: string, data: any) {
        const {offset, segmentNumber} = await this.segmentManager.set(key, data);
        // const {offset} = await this.fileAdapter.set(key, data)
        // Обновляем индекс: сохраняем смещение для ключа
        await this.indexManager.setOffset(key.toString(), offset, segmentNumber)
    }

    async get(key: string) {
        // Проверяем наличие ключа в индексе
        const indexData = await this.indexManager.getOffset(key);
        if (indexData === undefined) {
            return null; // Если ключа нет в индексе, возвращаем null
        }
        return await this.segmentManager.get(indexData.offset, indexData.segmentNumber);
    }

    async delete(key: string) {
        await this.segmentManager.delete(key);
        await this.indexManager.delete(key);

    }
}


export default SamuraiDB;